### PR TITLE
Correct trigger param parsing

### PR DESCRIPTION
**Pull request by SCdF**
If `params` doesn't parse right we were just dropping the error (yay callbacks!). I've refactored the code so that the error can be passed back through the main callback correctly.

First commit is just the tests, second is the code what fixes it.

image

### CODE REVIEW

File: transitions/registration.js [hunk 1]
```diff
@@ -218,36 +218,44 @@ module.exports = {
...
+                        // "{\"foo\": \"bar\"}"
+                        // And comma delimted strings:
+                        // "foo,bar" or just "foo"
+                        try {
+                            options.params = JSON.parse(event.params);
+                        } catch (e) {
+                            try {
+                                options.params = event.params.split(',');
+                            } catch (e) {
+                                return cb(new Error('event params cannot be parsed: must be either a comma-delimited String, or JSON *inside* a String'));
```
Comment (garethbowen) [inline transitions/registration.js]: "JSON inside a string" doesn't make sense to me. Can you say "serialized JSON" or is that too technical?

Comment (garethbowen) [inline transitions/registration.js]: Also, maybe include both `e`s in the error somewhere - they might be useful.

Comment (SCdF) [inline transitions/registration.js]: Yeah that was the least technical I could think to make that error on one line. However, whatever, let's make it really detailed with multiple lines and examples! It's the multi-line-string future!

**Comment by SCdF** [comment]
@garethbowen take another look please. I changed my approach a reasonable amount. The previous approach wouldn't actually detect *bad* JSON because it would just fall back to splitting a string.

So if you passed: `"{\"foo\": \"bar\""` (missing the closing `}`) it wouldn't actually error, because it would `JSON.parse()` that, fail, then `.split(',')` it, which would result in params equalling the entire string. Since params aren't required necessarily it would just happy process the doc and move on with its life.

This attempt does two things differently:
 - It supports mapping raw JSON. I'm not convinced our entire app will support this properly due to kanso.json expecting a string here, so it's still not our official method, but it's there anyway
 - We try to detect if your string is supposed to be JSON and just parse it that way, instead of falling back on string.split

**Pull request merged**